### PR TITLE
functional_tests: fix rpc payment tests spurious failure and speedup signature generation

### PR DESCRIPTION
**Pull request by moneromooo-monero**

### CODE REVIEW

File: tests/functional_tests/make_test_signature.cc [hunk 1]
```diff
@@ -54,7 +54,20 @@ int main(int argc, const char **argv)
     fprintf(stderr, "invalid secret key\n");
     return 1;
   }
-  std::string signature = cryptonote::make_rpc_payment_signature(skey);
-  printf("%s\n", signature.c_str());
+  uint32_t count = 1;
+  if (argc == 3)
+  {
+    count = atoi(argv[2]);
```
[the remark this answers is not in the document]
Comment (moneromooo-monero) [reply tests/functional_tests/make_test_signature.cc]: By this, do you mean "check for negative values" ? It's a test program so there's not much point. Or did you mean "use an int, not uint32_t" ?

**Comment by selsta** [comment]
Functional tests CI failed.
```
2020-03-20T17:48:59.1366737Z Testing access mining
2020-03-20T17:48:59.1366847Z Traceback (most recent call last):
2020-03-20T17:48:59.1366986Z   File "/home/runner/work/monero/monero/tests/functional_tests/rpc_payment.py", line 449, in <module>
2020-03-20T17:48:59.1367123Z     RPCPaymentTest().run_test()
2020-03-20T17:48:59.1367266Z   File "/home/runner/work/monero/monero/tests/functional_tests/rpc_payment.py", line 50, in run_test
2020-03-20T17:48:59.1367401Z     self.test_access_mining()
2020-03-20T17:48:59.1367542Z   File "/home/runner/work/monero/monero/tests/functional_tests/rpc_payment.py", line 295, in test_access_mining
2020-03-20T17:48:59.1368442Z     assert nonce < 1000 # can't find both valid and invalid -> the RPC probably fails
2020-03-20T17:48:59.1368580Z AssertionError
2020-03-20T17:48:59.1368688Z [TEST FAILED] rpc_payment
```

**Comment by moneromooo-monero** [comment]
Works for me.

**Comment by selsta** [comment]
Also works for me on locally but not in CI. CI does work consistently without this PR. Maybe someone else has an idea.

**Comment by moneromooo-monero** [comment]
Do you have daemon logs for this when it fails ?

**Comment by selsta** [comment]
This is the full log when running `make release-test`. Are you able to find the failing lines?

[bitmonero.log](https://github.com/monero-project/monero/files/4372758/bitmonero.log)

**Comment by moneromooo-monero** [comment]
No apparent error. Wallet logs maybe ?

**Comment by xiphon** [comment]
Think the reason is inconsistent `STALE_THRESHOLD`, it is 10s in `rpc_payment` test and 15s in the daemon code (`rpc_payment.cpp`).

```diff
diff --git "a/tests/functional_tests/rpc_payment.py" "b/tests/functional_tests/rpc_payment.py"
index 9a968d073..4d06feeb1 100644
--- "a/tests/functional_tests/rpc_payment.py"
+++ "b/tests/functional_tests/rpc_payment.py"
@@ -45,6 +45,7 @@ class RPCPaymentTest():
         assert len(self.make_test_signature) > 0
         self.secret_key, self.public_key = self.get_keys()
         self.signatures = []
+        self.stale_threshold = 15
         self.reset()
         self.test_access_tracking()
         self.test_access_mining()
@@ -170,7 +171,7 @@ class RPCPaymentTest():
             assert nonce < 1000 # can't find both valid and invalid -> the RPC probably fails
             last_credits = res.credits
 
-            if time.time() >= loop_time + 10:
+            if time.time() >= loop_time + self.stale_threshold:
                 res = daemon.rpc_access_info(client = self.get_signature())
                 cookie = res.cookie
                 loop_time = time.time()
@@ -206,7 +207,7 @@ class RPCPaymentTest():
             except:
                 found_invalid += 1
             assert nonce < 1000 # can't find a valid none -> the RPC probably fails
-            if time.time() >= loop_time + 10:
+            if time.time() >= loop_time + self.stale_threshold:
                 res = daemon.rpc_access_info(client = self.get_signature())
                 cookie = res.cookie
                 loop_time = time.time()
@@ -243,7 +244,7 @@ class RPCPaymentTest():
                 found_invalid += 1
             assert nonce < 1000 # can't find both valid and invalid -> the RPC probably fails
 
-            if time.time() >= loop_time + 10:
+            if time.time() >= loop_time + self.stale_threshold:
                 res = daemon.rpc_access_info(client = self.get_signature())
                 cookie = res.cookie
                 loop_time = time.time()
@@ -294,7 +295,7 @@ class RPCPaymentTest():
                     found_invalid += 1
             assert nonce < 1000 # can't find both valid and invalid -> the RPC probably fails
 
-            if time.time() >= loop_time + 10:
+            if time.time() >= loop_time + self.stale_threshold:
                 res = daemon.rpc_access_info(client = self.get_signature())
                 # cookie = res.cookie # let the daemon update its timestamp, but use old cookie
                 loop_time = time.time()
```

**Comment by selsta** [comment]
More logs (wallet logs, rpc logs,...):

[functional_tests.zip](https://github.com/monero-project/monero/files/4377730/functional_tests.zip)
[rpc-log.zip](https://github.com/monero-project/monero/files/4377731/rpc-log.zip)

**Comment by selsta** [comment]
@xiphon patch solves the issue.

**Comment by moneromooo-monero** [comment]
It's intended. The daemon rejects after 15 seconds, so the test refreshes after 10 to be sure not to be stale. If this actually fixes it, it would be nice to know why.

**Comment by selsta** [comment]
@moneromooo-monero these should be all the logs

[functional_tests.zip](https://github.com/monero-project/monero/files/4378048/functional_tests.zip)
[rpc-log.zip](https://github.com/monero-project/monero/files/4378049/rpc-log.zip)

**Comment by moneromooo-monero** [comment]
Does this help ?

```
diff --git a/tests/functional_tests/rpc_payment.py b/tests/functional_tests/rpc_payment.py
index 9a968d073..b15d7c3ca 100755
--- a/tests/functional_tests/rpc_payment.py
+++ b/tests/functional_tests/rpc_payment.py
@@ -280,6 +280,7 @@ class RPCPaymentTest():
         found_close_stale = 0
         found_late_stale = 0
         loop_time = time.time()
+        time.sleep(15)
         while found_close_stale == 0 or found_late_stale == 0:
             nonce += 1
             try:
```

**Pull request merged**